velox/expression/tests/PeeledEncodingTest.cpp

```diff
@@ -0,0 +1,408 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/PeeledEncoding.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::test;
+
+class PeeledEncodingTest : public testing::Test, public VectorTestBase {
+ public:
+  struct DictionaryWrap {
+    BufferPtr indices;
+    BufferPtr nulls;
+    int size;
+  };
+
+  static DictionaryWrap generateDictionaryWrap(
+      VectorFuzzer& fuzzer,
+      vector_size_t size) {
+    return {fuzzer.fuzzIndices(size, size), fuzzer.fuzzNulls(size), size};
+  }
+
+  // Utility function to wrap an arbitrary number of dictionary wrappings. Takes
+  // the 'base' vector to wrap, the number of wrappings, followed by an
+  // equivalent number of pointers to DictionaryWrap objects.
+  VectorPtr wrapInDictionaryLayers(VectorPtr base, int wrappings, ...) {
+    va_list args;
+    va_start(args, wrappings);
+    for (int i = 0; i < wrappings; i++) {
+      DictionaryWrap* wrap = va_arg(args, DictionaryWrap*);
+      base = BaseVector::wrapInDictionary(
+          wrap->nulls, wrap->indices, wrap->size, base);
+    }
+    va_end(args);
+    return base;
+  }
+
+  // Utility function that peels a given 'wrappedVector'. Takes the number of
+  // wraps to peel and a vector which is expected to have that many number of
+  // wraps.
+  VectorPtr peelWrappings(int numWrappingsToPeel, VectorPtr wrappedVector) {
+    if (numWrappingsToPeel == 0 ||
+        !PeeledEncoding::isPeelable(wrappedVector->encoding())) {
+      return wrappedVector;
+    } else {
+      return peelWrappings(
+          numWrappingsToPeel - 1, wrappedVector->valueVector());
+    }
+  }
+
+ protected:
+  core::ExecCtx execCtx_{pool_.get(), nullptr};
+};
+
+TEST_F(PeeledEncodingTest, basic) {
+  VectorFuzzer::Options options;
+  options.nullRatio = 0.3;
+  VectorFuzzer fuzzer(options, pool());
+  vector_size_t size = 100;
+
+  auto dictWrap1 = generateDictionaryWrap(fuzzer, size);
+  auto dictWrap2 = generateDictionaryWrap(fuzzer, size);
+  auto dictWrap3 = generateDictionaryWrap(fuzzer, size);
+
+  auto flat1 = fuzzer.fuzzFlat(INTEGER());
+  auto flat2 = fuzzer.fuzzFlat(INTEGER());
+  auto const1 = fuzzer.fuzzConstant(INTEGER());
+  auto const2 = fuzzer.fuzzConstant(INTEGER());
+  auto complexConst = fuzzer.fuzzConstant(ARRAY(INTEGER()));
+
+  SelectivityVector rows(size);
+  LocalDecodedVector localDecodedVector(execCtx_);
+
+  // 1. Common dictionary layers are peeled
```

Comment:
it would probably be nicer to break this large test into one test case per item you have, so in case something fails it's a bit easier to pinpoint but just looking at the test name. You could move the common data structures to a common parent class/object.

Reply:
done. Added parameterized gtests to account for variation in input selectivity vectors as well. Now a test failure should spit out all info required to pinpoint exactly which test failed and what kind of selectivity vector it was using.